File: ctrlfile/src/ast.zig
const std = @import("std");

pub const ActionKind = enum { cmd, install };

pub const Action = struct {
    kind: ActionKind,
    value: []const u8,
};

pub const Step = struct {
    number: u32,
    actions: []Action,
};

pub const RunBlock = struct {
    name: []const u8,
    needs: [][]const u8,
    steps: []Step,
};

pub const ExprKind = enum {
    string,
    call,
    eq,
    neq,
    and_,
    or_,
    not_,
    ternary,
};

pub const Expr = struct {
    kind: ExprKind,
    str: []const u8 = "",
    call_name: []const u8 = "",
    call_arg: []const u8 = "",
    left: ?*Expr = null,
    right: ?*Expr = null,
    cond: ?*Expr = null,
    then_expr: ?*Expr = null,
    else_expr: ?*Expr = null,
};

pub const WhenBlock = struct {
    condition: *Expr,
    actions: []Action,
};

pub const Program = struct {
    runs: []RunBlock,
    whens: []WhenBlock,
};

File: ctrlfilesafe/src/ast.rs
#[derive(Debug, Clone)]
pub enum Action {
    Cmd(Vec<String>),
    Shell(String),
    Install(String),
}

#[derive(Debug, Clone)]
pub struct Step {
    pub number: u32,
    pub actions: Vec<Action>,
}

#[derive(Debug, Clone)]
pub struct RunBlock {
    pub name: String,
    pub needs: Vec<String>,
    pub steps: Vec<Step>,
}

#[derive(Debug, Clone)]
pub enum Expr {
    Str(String),
    Call { name: String, arg: String },
    Eq(Box<Expr>, Box<Expr>),
    Neq(Box<Expr>, Box<Expr>),
    And(Box<Expr>, Box<Expr>),
    Or(Box<Expr>, Box<Expr>),
    Not(Box<Expr>),
    Ternary(Box<Expr>, Box<Expr>, Box<Expr>),
}

#[derive(Debug, Clone)]
pub struct WhenBlock {
    pub condition: Expr,
    pub actions: Vec<Action>,
}

#[derive(Debug, Clone, Default)]
pub struct Program {
    pub runs: Vec<RunBlock>,
    pub whens: Vec<WhenBlock>,
}
